### ISSUE TITLE
Named parameters and ParamArray do not not play nice together

### ISSUE THREAD

**Issue by jrmoreno1** [dotnet/vblang#597]
I'm creating an extension method 3 parameters, the last being a paramaray.  When attempting to reorder the parameters using named parameters,  the signature is unable to be matched.

```vb.net
Public Module MainModule 
    Sub Main()
      Dim b = "".Example("", i:=5)
    End Sub 

    <Extension>
    Public Function Example(sAs String, i As Integer, ParamArray p As String()) As Boolean
        Return False
    End Function
End Module 
```

This gets two errors, BC30512 and BC36584.

**Comment by RevensofT** [comment]
Isn't it always has been like this ? `ParamArray` mostly input unknow length of argument and it must be the last set of argument.

**Comment by zspitz** [comment]
Confirming what @RevensofT is saying, that arguments passed into `ParamArray` must always be last, this also doesn't compile:

```vb
Example("abcd", i:=5)

Function Example(i As Integer, ParamArray p As String()) As Boolean
    Return False
End Function
```

**Comment by jrmoreno1** [comment]
@zspitz : yes, they have to be last, that is what I meant by not playing nicely together.  The reason why ParmaArrays have to be last is that there is no other way of deciding when the parameters end UNLESS named arguments are used, in which case, all named arguments could be removed from consideration.

Before named parameters, given a call to fn(integer, paramarray integer,  integer, integer) it would have been impossible to determine whether fn(1,2,3) was supposed to represent fn(1, {}, 2,3) with an empty  parameter array, or whether it was supposed to represent fn(1, { 2}, 3)  where the last parameter was left off by mistake, or even fn(1, { 2,3}) and the last 2 were left off by mistake , with named parameters, the paramarray still needs to be **declared** last, eg sub fn(first as integer, second as integer, third as integer, paramarray i() as integer), but **calling** it as fn(4,5,6,third:=3, second:=2, first:=1) is no longer ambiguous. 

With named parameters, paramarray parameters should be more flexible.

**Comment by gilfusion** [comment]
Correct me if I'm wrong (only verified with a quick LINQPad check), but I think that the rule is that any unnamed parameters still need to be in order. For example, if the first parameter in the call is unnamed, it needs to be the actual first parameter. If you use a named version of the first parameter later, it's marked as an error because you're assigning the same parameter twice, and this behavior is not exclusive to methods with ParamArray.

**Comment by jrmoreno1** [comment]
@gilfusion : you can’t use a named parameter with param array.
``vb.net
  Dim b =  Example("", i:=5, p:= {"param1"})
``

which has all of elements in the correct order with the correct types, will fail, because you can’t use a named parameter with paramarray.

**Comment by jrmoreno1** [comment]
@Halofreak1990 : c# allows you to use named parameters and even reorder them, but isn’t smart enough to figure out that it can determine that the remaining items must be part of the params , by eliminating the other parameters.
